File: server/app/Models/ShipmentOffer.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;

class ShipmentOffer extends Model
{
    use SoftDeletes;

    /**
     * status values: pending, awaiting_manual_price, accepted, expired,
     * cancelled. Kept as a plain string column (no DB enum) so this new
     * value needed no schema change — see the 2026_08_15_000109 migration.
     */
    const STATUS_AWAITING_MANUAL_PRICE = 'awaiting_manual_price';

    protected $fillable = [
        'company_id',
        'tracking_number',
        'origin',
        'origin_lat',
        'origin_lng',
        'origin_country',
        'origin_city',
        'origin_zone_id',
        'origin_address',
        'destination',
        'destination_lat',
        'destination_lng',
        'destination_country',
        'destination_city',
        'destination_zone_id',
        'destination_address',
        'weight',
        'description',
        'needs_permit',
        'is_hazardous',
        'is_fragile',
        'order_type',
        'required_truck_type',
        'price_to_driver',
        'price_to_client',
        'pricing_mode',
        'priced_by_user_id',
        'platform_margin_percent_snapshot',
        // Zones / Smart Pricing Engine snapshot (2026-08-27) — see
        // 2026_08_27_000003_add_zone_pricing_fields_to_shipment_offers_table.php
        'pricing_reference',
        'pricing_low',
        'pricing_high',
        'pricing_confidence',
        'market_adjustment_snapshot',
        'pricing_level',
        'company_selected_price',
        'status',
        'financial_status',
        'cancellation_reason',
        'cancelled_by_user_id',
        'cancelled_at',
        'accepted_by_driver_id',
        'accepted_truck_id',
        'accepted_at',
        'expires_at',
        'matched_driver_ids',
        'matching_round',
    ];

    protected $casts = [
        'needs_permit' => 'boolean',
        'is_hazardous' => 'boolean',
        'is_fragile' => 'boolean',
        'accepted_at' => 'datetime',
        'expires_at' => 'datetime',
        'cancelled_at' => 'datetime',
        'platform_margin_percent_snapshot' => 'decimal:2',
        'origin_lat' => 'decimal:7',
        'origin_lng' => 'decimal:7',
        'destination_lat' => 'decimal:7',
        'destination_lng' => 'decimal:7',
        'matched_driver_ids' => 'array',
        'pricing_reference' => 'decimal:2',
        'pricing_low' => 'decimal:2',
        'pricing_high' => 'decimal:2',
        'market_adjustment_snapshot' => 'decimal:2',
        'company_selected_price' => 'decimal:2',
    ];

    public function company()
    {
        return $this->belongsTo(Company::class);
    }

    public function originZone()
    {
        return $this->belongsTo(Zone::class, 'origin_zone_id');
    }

    public function destinationZone()
    {
        return $this->belongsTo(Zone::class, 'destination_zone_id');
    }

    public function acceptedByDriver()
    {
        return $this->belongsTo(Driver::class, 'accepted_by_driver_id');
    }

    public function acceptedTruck()
    {
        return $this->belongsTo(Truck::class, 'accepted_truck_id');
    }

    public function pricedBy()
    {
        return $this->belongsTo(User::class, 'priced_by_user_id');
    }

    public function cancelledBy()
    {
        return $this->belongsTo(User::class, 'cancelled_by_user_id');
    }

    public function shipment()
    {
        return $this->hasOne(Shipment::class);
    }

    /** @return \Illuminate\Database\Eloquent\Relations\HasMany<ShipmentOfferMatchingRound> */
    public function matchingRounds()
    {
        return $this->hasMany(ShipmentOfferMatchingRound::class)->orderBy('round_number');
    }
}

File: app/lib/models/ShipmentOffer.dart
class ShipmentOffer {
  final int id;
  final int companyId;
  final String companyName;
  final String origin;
  final double? originLat;
  final double? originLng;
  final String destination;
  // Zones / Smart Pricing Engine (2026-08-27) — structured route fields,
  // present on both the company- and driver-facing JSON once an offer was
  // created through the zone-based flow; null for older/legacy offers.
  // pricing* fields are only ever present on the company-facing resource —
  // DriverFacingShipmentOfferResource deliberately omits them (a driver
  // never sees historical pricing/range/confidence), so they simply parse
  // as null for a driver's own offer list.
  final String? originCountry;
  final String? originCity;
  final int? originZoneId;
  final String? originAddress;
  final String? destinationCountry;
  final String? destinationCity;
  final int? destinationZoneId;
  final String? destinationAddress;
  final double? pricingReference;
  final double? pricingLow;
  final double? pricingHigh;
  final String? pricingConfidence;
  final String? pricingLevel;
  final String weight;
  final String description;
  final bool needsPermit;
  final bool isHazardous;
  final bool isFragile;
  final String orderType; // 'internal' or 'external'
  final String requiredTruckType;
  final String priceToDriver;
  final String priceToClient;
  final String? pricingMode; // 'auto', 'manual', or null
  // pending, awaiting_manual_price, escalated, accepted, cancelled
  final String status;
  final int eligibleDriversCount;
  final String createdAt;

  ShipmentOffer({
    required this.id,
    required this.companyId,
    required this.companyName,
    required this.origin,
    this.originLat,
    this.originLng,
    this.originCountry,
    this.originCity,
    this.originZoneId,
    this.originAddress,
    this.destinationCountry,
    this.destinationCity,
    this.destinationZoneId,
    this.destinationAddress,
    this.pricingReference,
    this.pricingLow,
    this.pricingHigh,
    this.pricingConfidence,
    this.pricingLevel,
    required this.destination,
    required this.weight,
    required this.description,
    required this.needsPermit,
    required this.isHazardous,
    required this.isFragile,
    required this.orderType,
    required this.requiredTruckType,
    required this.priceToDriver,
    required this.priceToClient,
    required this.pricingMode,
    required this.status,
    required this.eligibleDriversCount,
    required this.createdAt,
  });

  bool get isAwaitingManualPrice => status == 'awaiting_manual_price';
  bool get isEscalated => status == 'escalated';
  bool get isPending => status == 'pending';
  bool get hasZoneRoute => originZoneId != null && destinationZoneId != null;

  factory ShipmentOffer.fromJson(Map<String, dynamic> json) {
    bool asBool(dynamic v) => v == true || v == 1 || v == '1';

    return ShipmentOffer(
      id: json['id'] ?? 0,
      companyId: json['company_id'] ?? 0,
      companyName: json['company']?['name']?.toString() ?? '',
      origin: json['origin']?.toString() ?? '',
      originLat: json['origin_lat'] != null ? double.tryParse(json['origin_lat'].toString()) : null,
      originLng: json['origin_lng'] != null ? double.tryParse(json['origin_lng'].toString()) : null,
      originCountry: json['origin_country']?.toString(),
      originCity: json['origin_city']?.toString(),
      originZoneId: json['origin_zone_id'] is int ? json['origin_zone_id'] : int.tryParse(json['origin_zone_id']?.toString() ?? ''),
      originAddress: json['origin_address']?.toString(),
      destinationCountry: json['destination_country']?.toString(),
      destinationCity: json['destination_city']?.toString(),
      destinationZoneId: json['destination_zone_id'] is int ? json['destination_zone_id'] : int.tryParse(json['destination_zone_id']?.toString() ?? ''),
      destinationAddress: json['destination_address']?.toString(),
      pricingReference: json['pricing_reference'] != null ? double.tryParse(json['pricing_reference'].toString()) : null,
      pricingLow: json['pricing_low'] != null ? double.tryParse(json['pricing_low'].toString()) : null,
      pricingHigh: json['pricing_high'] != null ? double.tryParse(json['pricing_high'].toString()) : null,
      pricingConfidence: json['pricing_confidence']?.toString(),
      pricingLevel: json['pricing_level']?.toString(),
      destination: json['destination']?.toString() ?? '',
      weight: json['weight']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      needsPermit: asBool(json['needs_permit']),
      isHazardous: asBool(json['is_hazardous']),
      isFragile: asBool(json['is_fragile']),
      orderType: json['order_type']?.toString() ?? 'internal',
      requiredTruckType: json['required_truck_type']?.toString() ?? '',
      priceToDriver: json['price_to_driver']?.toString() ?? '',
      priceToClient: json['price_to_client']?.toString() ?? '',
      pricingMode: json['pricing_mode']?.toString(),
      status: json['status']?.toString() ?? 'pending',
      eligibleDriversCount: json['eligible_drivers_count'] is int
          ? json['eligible_drivers_count']
          : int.tryParse(json['eligible_drivers_count']?.toString() ?? '') ??
              0,
      createdAt: json['created_at']?.toString() ?? '',
    );
  }
}
